src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -188,7 +195,7 @@ private void EnsureStartup()
                 return;
             }
 
-            _startup = _hostingServiceProvider.GetService<IStartup>();
+            _startup = (_applicationServices ?? _hostingServiceProvider).GetService<IStartup>();
```

Comment:
I am not sure if resolved IStartup instance should be registered with _hostingServiceProvider as well. If it does, let me know. I'll add it.

Reply:
Why are you resolving the IStartup from the application services? That's just broken. The hosting service provider is responsible for building the IStartup. There are 2 separate phases.

Reply:
At this point in time ``_applicationServices`` holds reference to externally supplied ``IServiceProvider`` or ``null`` in which case normal flow kicks in and ``IStartup`` is resolved from ``_hostingServiceProvider``.